python/ray/includes/unique_ids.pxi

```diff
@@ -326,8 +329,19 @@ cdef class ClientActorRef(ActorID):
             # call_release in this case, since the client should have already
             # disconnected at this point.
             return
-        if client.ray.is_connected() and not self.data.IsNil():
-            client.ray.call_release(self.id)
+        if client.ray.is_connected():
+            try:
+                self._wait_for_id(timeout=10)
```

Comment:
The question here: any reason you pick 10s? Another question is that if it failed to be resolved, will it lead to a leak?

Reply:
The timeout shouldn't be too long because it is in destructor, and the user doesn't need the ref / object anymore. We can lower the timeout to 5 or 2. Timeout for `ray.get()` is 2s.

When there is a failure to resolve an ID, the ID will not be released so the ref will be kept server-side after the client ref is gone. Object and actor refs are still cleaned up server-side, when client-server shuts down. Failure to resolve an ID most likely happens after a failed connection or server. So this should not be a significant issue. Before this change, this issue can happen in theory if a ClientTaskTicket response is lost, although this is more rare.

Reply:
I feel this is somewhat still important. It's still a leak and under unstable internet environment it's going to happen a lot. Although it will be released once the client exits, it still leaks in the code and that program will fail.

Btw, how is the reconnect going to work here? Will the failed resolving be called again?

And is it better to retry this and fail the program in the extreme case (for example failed to do cleanup for 5 times)

Or should we make the timeout here to be infinite?

Reply:
When the resolving takes too long and fails, if the resolution succeeds in the background with a server response, the ID future will be set but there will be not ref counting. I thought about it for a bit, and it seems there is no issue with using an infinite timeout here. It becomes closer to the behavior before this PR, where client waits synchronously for response from server.